pydis_site/apps/home/tests/test_signal_listener.py

```diff
@@ -0,0 +1,392 @@
+        )
+
+        cls.discord_unmapped.roles.add(cls.unmapped_role)
+        cls.discord_unmapped.save()
+
+        cls.discord_not_in_guild = DiscordUser.objects.create(
+            id=3,
+            name="not-in-guild",
+            discriminator=0,
+            avatar_hash=None,
+            in_guild=False
+        )
+
+        cls.discord_admin = DiscordUser.objects.create(
+            id=1,
+            name="admin",
+            discriminator=0,
+            avatar_hash=None
+        )
+
+        cls.discord_admin.roles.set([cls.admin_role])
+        cls.discord_admin.save()
+
+        cls.discord_moderator = DiscordUser.objects.create(
+            id=4,
+            name="admin",
+            discriminator=0,
+            avatar_hash=None
+        )
+
+        cls.discord_moderator.roles.set([cls.moderator_role])
+        cls.discord_moderator.save()
+
+        cls.django_user_discordless = DjangoUser.objects.create(username="no-discord")
+        cls.django_user_never_joined = DjangoUser.objects.create(username="never-joined")
+
+        cls.social_never_joined = SocialAccount.objects.create(
+            user=cls.django_user_never_joined,
+            provider=DiscordProvider.id,
+            uid=5
+        )
+
+        cls.django_user = DjangoUser.objects.create(username="user")
+
+        cls.social_user = SocialAccount.objects.create(
+            user=cls.django_user,
+            provider=DiscordProvider.id,
+            uid=cls.discord_user.id
+        )
+
+        cls.social_user_github = SocialAccount.objects.create(
+            user=cls.django_user,
+            provider=GitHubProvider.id,
+            uid=cls.discord_user.id
+        )
+
+        cls.social_unmapped = SocialAccount(
+            # We instantiate it and don't put it in the DB. This is (surprisingly)
+            # a realistic test case, so we need to check for it
+
+            provider=DiscordProvider.id,
+            uid=5,
+            user_id=None  # No relation exists at all
+        )
+
+        cls.django_admin = DjangoUser.objects.create(
+            username="admin",
+            is_staff=True,
+            is_superuser=True
+        )
+
+        cls.social_admin = SocialAccount.objects.create(
+            user=cls.django_admin,
+            provider=DiscordProvider.id,
+            uid=cls.discord_admin.id
+        )
+
+        cls.django_moderator = DjangoUser.objects.create(
+            username="moderator",
+            is_staff=True,
+            is_superuser=False
+        )
+
+        cls.social_moderator = SocialAccount.objects.create(
+            user=cls.django_moderator,
+            provider=DiscordProvider.id,
+            uid=cls.discord_moderator.id
+        )
+
+    def test_model_save(self):
+        """Test signal handling for when Discord user model objects are saved to DB."""
+        mock_obj = mock.Mock()
+
+        with mock.patch.object(SignalListener, "_apply_groups", mock_obj):
+            _ = SignalListener()
```

Comment:
Would it work to just not store this at all? like just doing `SignalListener()`? I'm mostly curious, the way you've done it here doesn't really bother me, but it might be cleaner to just call it like a function if it actually works.

Reply:
I kind of just pre-emptively did this to make sure garbage collection didn't screw us over, but I can check.